kubelet: add support for broadcasting metrics from CRI

haircommander opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature
#### What this PR does / why we need it:
This PR includes https://github.com/kubernetes/kubernetes/pull/113025 as well as support in the kubelet for pulling the metrics from CRI and broadcasting them to prometheus

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:
There are two open questions that I need someone who knows more about the way these endpoints work:
- [ ] should the collecting be async for each metric
- [x] is it idiomatic to return the key and value of labels as a map pair/struct. I see most examples as having constant labels, and others as building each slice each time, but it seems inefficient to me

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add alpha support for returning container and pod metrics from CRI, instead of cAdvsior
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:

- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/2371
```

danielye11 committed:
Add fake runtimes and CRI changes for KEP-2371

Added new gRPC call 'ListPodSanboxMetrics' which would return additional
container stats currently supported by cAdvisor, but outside the scope
of /stats/summary api. Added new types to support metric exporting of
prometheus, including Metric and other subfields. Added fake runtime
changes associated with the CRI changes.

haircommander committed:
kubelet: wire ListPodSandboxMetrics

Signed-off-by: Peter Hunt <pehunt@redhat.com>

haircommander committed:
kubelet/metrics: add cri_metrics

that pulls metrics from the CRI

Signed-off-by: Peter Hunt <pehunt@redhat.com>

haircommander committed:
kubelet: add cri metrics to server

Signed-off-by: Peter Hunt <pehunt@redhat.com>

haircommander commented:
/sig node
/priority important-soon

bobbypage commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1692,3 +1695,40 @@ enum ContainerEventType {
+
+message Metric {
+    string name = 1;
+    string help = 2;
```
are we sure we want help here? This is a bit unfortunate because it means we need to send help text for every single metric when help is the same for the same metric "name". It seems like it will increase the rpc payload size alot due to the duplication (since the metric is per container). 

Maybe we can omit this for now and figure out best way to include this later?

bobbypage commented on pkg/kubelet/metrics/collectors/cri_metrics.go:
```diff
@@ -0,0 +1,115 @@
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
```
I think we need to move this to `k8s.io/component-base/metrics` based on discussion I had with @dashpole. 

Can we use that + `NewLazyConstMetric` - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/value.go#L43? Or is something preventing us?

haircommander replied:
the problem is if the CRI implementation messes up the metric, it causes kubelet to panic. I tried that originally, but in my fussing with the metrics, I kept causing kubelet to crash. NewLazyConstMetric seems to be more in situations we have the metrics/descriptions precalculated and are willing to panic if they're wrong (programmer error vs runtime error)

dashpole replied:
That is pretty concerning...

If we need it here, we would need to add NewConstMetric (without Must) to component-base, rather than use the upstream prometheus library.  cc @logicalhan

haircommander replied:
The other reason I went for this is there isn't really a notion of stability with these metrics (unless we wire them through the cri, or assume they have the stability of the KEP)

would you prefer I:
- add NewConstMetric function
- use MustNewConstMetric despite the risk
- continue this way until we have a more clear path forward?

From my perspective, cadvisor is currently registered with RawMustRegister, so this way is the most in-line with cadvisor's behavior. I feel like it could be a follow-up to update

dashpole replied:
I would prefer:

* add NewConstMetric function

We should either:

* Mark all metrics as Internal stability, or
* Add stability to the CRI API, and use that.

bobbypage commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1692,3 +1695,40 @@ enum ContainerEventType {
+
+message Metric {
+    string name = 1;
+    string help = 2;
```
are we sure we want help here? This is a bit unfortunate because it means we need to send help text for every single metric when help is the same for the same metric "name". It seems like it will increase the rpc payload size alot due to the duplication (since the metric is per container). 

Maybe we can omit this for now and figure out best way to include this later?

haircommander replied:
I wonder if we should dedup the values we're sending over. rather than defining the metric per metric, CRI could return a name -> metric (including everything but value), and then for each container/pod it could send a map of name and value. Would dedup a lot of the information we're sending (basically removing lots of duplicated metadata)

dashpole replied:
That does mean we are making this somewhat harder for CRI implementations to implement...  I do think help is important to include.

bobbypage commented on pkg/kubelet/metrics/collectors/cri_metrics.go:
```diff
@@ -0,0 +1,115 @@
+
+// Describe implements the prometheus.Describe interface.
+// TODO(haircommander): would it be better if these were processed async?
+func (c *criMetricsCollector) Describe(ch chan<- *prometheus.Desc) {
```
Do you know when Describe and when Collect are collect? If for every scrape describe gets called we end up having to make two calls to the CRI :( 

Maybe we should be following the pattern of making the descriptors on the fly similar to cAdvisor approach? - https://github.com/google/cadvisor/blob/master/metrics/prometheus.go#L1900

Though we can always optimize this later, happy to start with this approach and iterate.

haircommander replied:
we're currently creating them on the fly, we'd have to cache them if we wanted to save the value (which I'm open to, but would further complicate kubelet implementation and cause drift)

dashpole replied:
Describe is called once at the startup, IIRC.  You are allowed to not implement Describe, but it removes a lot of important validation.

LuBingtan reviewed (commented):
nice work :)

haircommander committed:
component-base/metrics: Add NewConstMetric function

so that a caller can use the metrics.Metric structure but still handle
errors

Signed-off-by: Peter Hunt <pehunt@redhat.com>

haircommander commented:
a couple of highlights to the updates:
- Separated CRI calls for ListMetricDescriptors and ListPodSandboxMetrics. That allows for there to be less overhead, as more information can be called once and then cached
- Added NewConstMetric call and moved to comp-base metrics instead of prometheus
    - I went for INTERNAL metric type to start (WDYT @bobbypage )
- Did a bit of cleanup here and there

PTAL @dashpole @bobbypage

mrunalp commented on pkg/kubelet/cri/remote/fake/fake_runtime.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2017 The Kubernetes Authors.
+Copyright 2015 The Kubernetes Authors.
```
Probably not needed?

haircommander replied:
boiler plate is messing with me

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1692,3 +1701,57 @@ enum ContainerEventType {
+	// labels that have previously been declared in label_keys.
+	// It is the responsibility of the runtime to correctly keep sorted the keys and values.
+	// If the two slices have different length, the behavior is undefined.
+	repeated string label_keys = 3;
```
Whitespace is off inside this message.

haircommander replied:
fixed

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1692,3 +1701,57 @@ enum ContainerEventType {
+	// The corresponding LabelValues to the LabelKeys defined in the MetricDescriptor.
+	// It is the responsibility of the runtime to correctly keep sorted the keys and values.
+	// If the two slices have different length, the behavior is undefined.
+	repeated string label_values = 4;
```
whitespace off.

haircommander replied:
fixed

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1692,3 +1701,57 @@ enum ContainerEventType {
+
+enum MetricType {
+    COUNTER = 0;
+    GAUGE = 1;
```
Do we foresee adding more types?

haircommander replied:
the only other one the kubernetes metrics package supports is UNTYPED, so maybe? but it doesn't seem to be used in cadvisor rn and I'm not really clear on the usage

haircommander commented:
/retest

dashpole commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -798,3 +798,35 @@ func (r *remoteRuntimeService) CheckpointContainer(options *runtimeapi.Checkpoin
+}
+
+// ListPodSandboxMetrics returns the list of pod sandbox metrics, additional CPU, memory, process stats
+func (r *remoteRuntimeService) ListPodSandboxMetrics() ([]*runtimeapi.PodSandboxMetrics, error) {
+	// Set timeout, because runtimes are able to cache disk stats results
+	ctx, cancel := getContextWithTimeout(r.timeout)
+	defer cancel()
```
```suggestion
func (r *remoteRuntimeService) ListPodSandboxMetrics(ctx context.Context) ([]*runtimeapi.PodSandboxMetrics, error) {
	// Set timeout, because runtimes are able to cache disk stats results
	ctx, cancel := context.WithTimeout(ctx, r.timeout)
	defer cancel()
```
getContextWithTimeout is removed at HEAD, so you will need to switch off of this.  Please pass context to all added runtime functions.

dashpole replied:
See https://github.com/kubernetes/kubernetes/pull/113591

dashpole commented on staging/src/k8s.io/component-base/metrics/value.go:
```diff
@@ -47,6 +47,16 @@ func NewLazyConstMetric(desc *Desc, valueType ValueType, value float64, labelVal
+// NewConstMetric is a helper of NewConstMetric.
+//
+// Note: If the metrics described by the desc is hidden, the metrics will not be created.
+func NewConstMetric(desc *Desc, valueType ValueType, value float64, labelValues ...string) (Metric, error) {
```
FYI @logicalhan

dashpole commented on pkg/kubelet/metrics/collectors/cri_metrics.go:
```diff
@@ -0,0 +1,118 @@
+// NewCRIMetricsCollector implements the metrics.Collector interface
+func NewCRIMetricsCollector(ctx context.Context, listPodSandboxMetrics func(context.Context) ([]*runtimeapi.PodSandboxMetrics, error), listMetricDescriptors func(context.Context) ([]*runtimeapi.MetricDescriptor, error)) metrics.StableCollector {
+	descs, err := listMetricDescriptors(ctx)
+	if err != nil {
```
Does this mean that if we fail the initial descriptor list, we never get any help text?

haircommander replied:
yeah. I suppose we could re-request if we see the structure is empty, though I think we'd also have to add some locking to do so.

haircommander replied:
thinking about it more, I think for now this is sufficient. Since they're separated calls, we can always add additional calls in the future, but this is conceptually simple and avoids excessive work when CRI doesn't support the call

haircommander commented:
/retest

haircommander commented:
tests happy and comments addressed

PTAL @mrunalp @dashpole @bobbypage

haircommander commented:
updated with Timestamp handling (thanks @rphillips for pointing it out)

bobbypage commented on pkg/kubelet/metrics/collectors/cri_metrics.go:
```diff
@@ -0,0 +1,123 @@
+type criMetricsCollector struct {
+	metrics.BaseStableCollector
+	descriptors           map[string]*metrics.Desc
+	listPodSandboxMetrics func(context.Context) ([]*runtimeapi.PodSandboxMetrics, error)
```
nit: maybe rename to listPodSandboxMetricsFn to make it clear this is a function?

bobbypage commented on pkg/kubelet/metrics/collectors/cri_metrics.go:
```diff
@@ -0,0 +1,123 @@
+}
+
+func criDescToProm(m *runtimeapi.MetricDescriptor) *metrics.Desc {
+	return metrics.NewDesc(m.Name, m.Help, m.LabelKeys, nil, metrics.INTERNAL, "")
```
nit: maybe worth to note here, the label keys are "variableLabels" as compared to "constant labels" (which is passed as nil), because the values of the labels will be different for each container (xref: https://github.com/kubernetes/kubernetes/blob/master/vendor/github.com/prometheus/client_golang/prometheus/desc.go#L77-L81)

bobbypage commented on pkg/kubelet/metrics/collectors/cri_metrics.go:
```diff
@@ -0,0 +1,123 @@
+func (c *criMetricsCollector) criMetricToProm(m *runtimeapi.Metric) (metrics.Metric, error) {
+	desc, ok := c.descriptors[m.Name]
+	if !ok {
+		err := fmt.Errorf("Error converting CRI Metric to prometheus format")
```
nit: avoid capatilizing error (https://github.com/golang/go/wiki/CodeReviewComments#error-strings)

bobbypage commented on pkg/kubelet/metrics/collectors/cri_metrics.go:
```diff
@@ -0,0 +1,123 @@
+
+	pm, err := metrics.NewConstMetric(desc, typ, float64(m.GetValue().Value), m.LabelValues...)
+	if err != nil {
+		klog.ErrorS(err, "Error getting summary for resourceMetric prometheus endpoint")
```
If there is an error here, it means that prometheus was unable to unable intilize the constant metric (i.e. `prometheus.NewConstMetric` failed) right? Might be worth to include some more info then since it likely something may be wrong with the descriptor / values?: Maybe we can use the existing `desc.String()` function - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/desc.go#L91-L97

```
klog.ErrorS(err, "Error creating CRI prometheus metric", "metric", desc.String())
```

bobbypage commented on pkg/kubelet/metrics/collectors/cri_metrics.go:
```diff
@@ -0,0 +1,123 @@
+		klog.ErrorS(err, "Error getting summary for resourceMetric prometheus endpoint")
+		return nil, err
+	}
+	// If Timestamp is 0, then the runtime did not cache the result.
```
I'm still a bit confused on what we mean by runtime caching the metric (see other related comment).

bobbypage commented on staging/src/k8s.io/cri-api/pkg/apis/testing/fake_runtime_service.go:
```diff
@@ -64,11 +64,14 @@ type FakeRuntimeService struct {
+	FakePodSandboxStats     map[string]*runtimeapi.PodSandboxStats
+	FakePodSandboxMetrics   map[string]*runtimeapi.PodSandboxMetrics
+	FakeMetricDescriptors   map[string]*runtimeapi.MetricDescriptor
+	FakePodContainerMetrics map[string]*runtimeapi.ContainerMetrics
```
Maybe rename to `FakeContainerMetrics`?

bobbypage commented:
Left a few small nits, but overall looks good! Also a few questions about what "timeStamp=0" / what do we mean by runtime caching a metric.

dashpole commented:
/approve

haircommander commented on pkg/kubelet/cri/remote/fake/fake_runtime.go:
```diff
@@ -336,3 +336,23 @@ func (f *RemoteRuntime) CheckpointContainer(ctx context.Context, req *kubeapi.Ch
 	return nil
 }
+
+// ListMetricDescriptors returns metrics of all running pods, additional CPU, memory, process stats
```
oop copypaste good catch

haircommander commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1702,3 +1711,57 @@ enum ContainerEventType {
+    // Name must match a name previously returned in a MetricDescriptors call,
+    // otherwise, it will be ignored.
+    string name = 1;
+    // Return 0 if metric is cached
```
right this comment is inverted .I'll update

haircommander commented:
thanks for the review @bobbypage, comments addressed

derekwaynecarr commented on pkg/kubelet/server/server.go:
```diff
@@ -400,14 +399,19 @@ func (s *Server) InstallDefaultHandlers() {
+	// cAdvisor metrics are exposed under the secured handler as well
+	r := compbasemetrics.NewKubeRegistry()
 	r.RawMustRegister(metrics.NewPrometheusMachineCollector(prometheusHostAdapter{s.host}, includedMetrics))
+	if utilfeature.DefaultFeatureGate.Enabled(features.PodAndContainerStatsFromCRI) {
```
this looks sufficiently isolated to make it safe for alpha.

all the code falls back to existing behavior when not enabled.

derekwaynecarr reviewed (approved):
/approve
/lgtm

we need to measure performance before enabling beta.

dims commented:
/triage accepted

bobbypage commented:
Thank you for the updates!

/lgtm

dims commented:
re-apply LGTM

/lgtm

bobbypage commented:
/lgtm

haircommander commented:
/retest
